### PR TITLE
[CloudPC]Add restore and bulkRestore API 

### PR DESCRIPTION
**Pull request by rongting**

### CODE REVIEW

**Comment by RuiHou105** [comment]
@Lauragra  @JarbasHorst @isvargasmsft  @Danielabom could you please help review this PR as the graph schema already went alive? Thanks ahead!

**Comment by JarbasHorst** [comment]
Sure, @RuiHou105! I will review this PR.

File: api-reference/beta/resources/cloudpcbulkremoteactionresult.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Represents cloud PC bulk remote action result.
+
+## Properties
+|Property|Type|Description|
+|:---|:---|:---|
+|successfulDeviceIds|String collection|A list of all the managed devices id that completed the bulk action successfully.|
```
Comment (JarbasHorst) [inline api-reference/beta/resources/cloudpcbulkremoteactionresult.md:21]: For clarity, please replace "managed devices id" with "Intune managed device IDs".

Please, apply this change also to lines 22, 23, and 24.

File: changelog/Microsoft.CloudManagedDesktop.json [hunk 2]
```diff
@@ -105,6 +105,56 @@
...
+                    "ChangeType": "Addition",
+                    "Description": "Added the **restoreTimeRange** enumeration type.",
+                    "Target": "restoreTimeRange"
+                },
+                {
+                    "Id": "d5711d39-7d15-4fac-9d6d-e81c2feae401",
+                    "ApiChange": "Resource",
+                    "ChangedApiName": "cloudPcBulkRemoteActionResult",
+                    "ChangeType": "Addition",
+                    "Description": "Added the [cloudPcBulkRemoteActionResult](https://docs.microsoft.com/en-us/graph/api/resources/intune-cloudPcBulkRemoteActionResult?view=graph-rest-beta) resource type.",
```
Comment (JarbasHorst) [inline changelog/Microsoft.CloudManagedDesktop.json:131]: The URL in the description is wrong => **intune-cloudPcBulkRemoteActionResult** should be **cloudPcBulkRemoteActionResult**.

File: changelog/Microsoft.CloudManagedDesktop.json [hunk 3]
```diff
@@ -105,6 +105,56 @@
...
+                    "Target": "cloudPcBulkRemoteActionResult"
+                },
+                {
+                    "Id": "d5711d39-7d15-4fac-9d6d-e81c2feae401",
+                    "ApiChange": "Method",
+                    "ChangedApiName": "bulkRestoreCloudPc",
+                    "ChangeType": "Addition",
+                    "Description": "Added the [bulkRestoreCloudPc](https://docs.microsoft.com/en-us/graph/api/intune-bulkRestoreCloudPc?view=graph-rest-beta) method to the [bulkRestoreCloudPc](https://docs.microsoft.com/en-us/graph/api/intune-bulkRestoreCloudPc?view=graph-rest-beta) resource.",
```
Comment (JarbasHorst) [inline changelog/Microsoft.CloudManagedDesktop.json:139]: Please, replace this description with:

- Added the \[bulkRestoreCloudPc\](https://docs.microsoft.com/en-us/graph/api/manageddevice-bulkrestorecloudpc?view=graph-rest-beta) method to the \[Cloud PC\](https://docs.microsoft.com/en-us/graph/api/cloudpc?view=graph-rest-beta) resource.

File: changelog/Microsoft.CloudManagedDesktop.json [hunk 4]
```diff
@@ -105,6 +105,56 @@
...
+                    "Target": "bulkRestoreCloudPc"
+                },
+                {
+                    "Id": "d5711d39-7d15-4fac-9d6d-e81c2feae401",
+                    "ApiChange": "Method",
+                    "ChangedApiName": "restoreCloudPc",
+                    "ChangeType": "Addition",
+                    "Description": "Added the [restoreCloudPc](https://docs.microsoft.com/en-us/graph/api/intune-restoreCloudPc?view=graph-rest-beta) method to the [restoreCloudPc](https://docs.microsoft.com/en-us/graph/api/intune-restoreCloudPc?view=graph-rest-beta) resource.",
```
Comment (JarbasHorst) [inline changelog/Microsoft.CloudManagedDesktop.json:147]: Please, replace this description with:

- Added the \[restoreCloudPc\](https://docs.microsoft.com/en-us/graph/api/manageddevice-restorecloudpc?view=graph-rest-beta) method to the \[Cloud PC\](https://docs.microsoft.com/en-us/graph/api/cloudpc?view=graph-rest-beta) resource.

File: changelog/Microsoft.CloudManagedDesktop.json [hunk 5]
```diff
@@ -105,6 +105,56 @@
...
+                    "Target": "restoreCloudPc"
```
Comment (JarbasHorst) [inline changelog/Microsoft.CloudManagedDesktop.json:148]: Please, replace **restoreCloudPc** with **managedDevice**.

File: changelog/Microsoft.CloudManagedDesktop.json [hunk 6]
```diff
@@ -105,6 +105,56 @@
...
+                    "Description": "Added the [cloudPcBulkRemoteActionResult](https://docs.microsoft.com/en-us/graph/api/resources/intune-cloudPcBulkRemoteActionResult?view=graph-rest-beta) resource type.",
+                    "Target": "cloudPcBulkRemoteActionResult"
+                },
+                {
+                    "Id": "d5711d39-7d15-4fac-9d6d-e81c2feae401",
+                    "ApiChange": "Method",
+                    "ChangedApiName": "bulkRestoreCloudPc",
+                    "ChangeType": "Addition",
+                    "Description": "Added the [bulkRestoreCloudPc](https://docs.microsoft.com/en-us/graph/api/intune-bulkRestoreCloudPc?view=graph-rest-beta) method to the [bulkRestoreCloudPc](https://docs.microsoft.com/en-us/graph/api/intune-bulkRestoreCloudPc?view=graph-rest-beta) resource.",
+                    "Target": "bulkRestoreCloudPc"
```
Comment (JarbasHorst) [inline changelog/Microsoft.CloudManagedDesktop.json:140]: Please, replace **bulkRestoreCloudPc** with **managedDevice**.

File: changelog/Microsoft.CloudManagedDesktop.json [hunk 7]
```diff
@@ -105,6 +105,56 @@
...
             "SubArea": "Cloud PC"
         },
+        {
+            "ChangeList": [
+                {
+                    "Id": "d5711d39-7d15-4fac-9d6d-e81c2feae401",
+                    "ApiChange": "Member",
+                    "ChangedApiName": "restoring",
+                    "ChangeType": "Addition",
+                    "Description": "Added the **restoring** member to the **cloudPcStatus** enumeration.",
```
Comment (JarbasHorst) [inline changelog/Microsoft.CloudManagedDesktop.json:115]: For consistency, please replace \*\*restoring\*\* with \`restoring\`.

File: api-reference/beta/resources/cloudpcbulkremoteactionresult.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+The following is a JSON representation of the resource.
+<!-- {
+  "blockType": "resource",
+  "@odata.type": "microsoft.graph.cloudPcBulkRemoteActionResult"
+}
+-->
+``` json
+{
+  "@odata.type": "#microsoft.graph.cloudPcBulkRemoteActionResult",
+  "successfulDeviceIds": [
```
Comment (JarbasHorst) [inline api-reference/beta/resources/cloudpcbulkremoteactionresult.md:39]: Please show the properties here in alphabetical order. You don't need to change the position of @odata…

File: api-reference/beta/api/manageddevice-bulkrestorecloudpc.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+## Examples
+
+### Request
+<!-- {
+  "blockType": "request",
+  "name": "manageddevicethis.bulkrestorecloudpc"
+}
+-->
+``` http
+POST https://graph.microsoft.com/beta/me/managedDevices/bulkRestoreCloudPc
```
Comment (RuiHou105) [inline api-reference/beta/api/manageddevice-bulkrestorecloudpc.md:75]: beta/deviceManagement/managedDevices/bulkRestoreCloudPc. Not "`me`"

File: api-reference/beta/api/manageddevice-bulkrestorecloudpc.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    "7c82a3e3-9459-44e4-94d9-b92f93bf78dd"
+  ],
+  "restorePointDateTime": "2021-09-23T04:00:00.0000000",
+  "timeRange": "before"
+}
+```
+
+
+### Response
+>**Note:** The response object shown here might be shortened for readability.
```
Comment (RuiHou105) [inline api-reference/beta/api/manageddevice-bulkrestorecloudpc.md:91]: @rongting  I suppose this note is extra and can be removed.

File: api-reference/beta/api/manageddevice-restorecloudpc.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+## Examples
+
+### Request
+<!-- {
+  "blockType": "request",
+  "name": "manageddevicethis.restorecloudpc"
+}
+-->
+``` http
+POST https://graph.microsoft.com/beta/me/managedDevices/{managedDeviceId}/restoreCloudPc
```
Comment (RuiHou105) [inline api-reference/beta/api/manageddevice-restorecloudpc.md:66]: beta/deviceManagement/managedDevices/bulkRestoreCloudPc. Not "`me`"

File: api-reference/beta/api/manageddevice-restorecloudpc.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+Content-length: 37
+
+{
+  "cloudPcSnapshotId": "A00009UV000_93aff428-61f2-467f-a879-1102af6fd4a8"
+}
+```
+
+
+### Response
+>**Note:** The response object shown here might be shortened for readability.
```
Comment (RuiHou105) [inline api-reference/beta/api/manageddevice-restorecloudpc.md:77]: I think this note can be removed.

File: api-reference/beta/resources/cloudpc.md [hunk 13]
```diff
@@ -31,6 +31,8 @@ Represents a cloud-managed virtual desktop. This Cloud PC is also enrolled into
 |[Bulk reprovision remote action](../api/manageddevice-bulkreprovisioncloudpc.md)|None|Bulk reprovision a set of Cloud PC devices with Intune managed device IDs.|
 |[Resize remote action](../api/manageddevice-resizecloudpc.md)|None|Upgrade or downgrade an existing Cloud PC to another configuration with new vCPU and storage size through Intune managed device ID.|
 |[Troubleshoot](../api/cloudpc-troubleshoot.md)|None|Troubleshoot a specific [cloudPC](../resources/cloudpc.md) object. Use this API to check the health status of the Cloud PC and the session host.|
+|[Restore remote action](../api/manageddevice-restorecloudpc.md)|None|Restore a Cloud PC device to a previous state from a snapshot.|
+|[Bulk restore remote action](../api/manageddevice-bulkrestorecloudpc.md)|[cloudPcBulkRemoteActionResult](../resources/cloudpcbulkremoteactionresult.md)|Restore multiple Cloud PC devices with a single request that includes Intune managed device IDs and a restore point date and time.|
```
Comment (JarbasHorst) [inline api-reference/beta/resources/cloudpc.md:35]: For clarity, please replace "Intune managed device IDs" with "the IDs of Intune managed devices" =>

- ..includes the IDs of Intune managed devices and a...

**Comment by JarbasHorst** [comment]
@RuiHou105, @rongting, checking to see if you have any questions regarding some of the feedback I provided?

**Comment by JarbasHorst** [comment]
Hi @rongting,

this PR isn't ready to merge yet. There are still seven remaining comments that you need to address first. Can you please check the open comments?

image

File: api-reference/beta/toc.yml [hunk 14]
```diff
@@ -10020,6 +10020,10 @@ items:
             href: api/manageddevice-bulkreprovisioncloudpc.md
           - name: Resize remote action
             href: api/manageddevice-resizecloudpc.md
+          - name: Restore remote action
```
Comment (Lauragra) [inline api-reference/beta/toc.yml:10065]: I'm not clear about why we're including "remote action" in the TOC entries when that isn't part of the descriptions for the methods. Can we shorten these entries to just "Restore" and "Bulk restore", or should we add "remotely" to the method descriptions?

**Comment by Lauragra** [comment]
@rongting , can you please see my comments/question regarding the TOC on this PR? When we resolve that, we'll be able to merge and publish. Thanks.

File: api-reference/beta/toc.yml (hunk 14 above)
Comment (AshleyYangSZ) [reply api-reference/beta/toc.yml]: I resolve this comment, @Lauragra please let me know if you still have concern about the name of these TOC entries.

Comment (AshleyYangSZ) [reply api-reference/beta/toc.yml]: This is a long story. When we first added reprovision API, which supports to perform on CloudPC and managedDevices, thus both /deviceManagement/virutalEndpoint/CloudPCs/Id/reprovision, and /deviceManagement/managedDevices/id/reprovisionCloudPc are supported. To distinguish the two API, we named the first one just as reprovision, the last one as reprovision remote action. So, I agree to update it to remotely action. Is it OK? @Lauragra

**Review by AshleyYangSZ — APPROVED** [review]

**Review by Lauragra — APPROVED** [review]

File: api-reference/beta/toc.yml (hunk 14 above)
Comment (Lauragra) [reply api-reference/beta/toc.yml]: @AshleyYangSZ , looks good to me now. Please set **ready to merge** when you want to publish, I see that the label has been removed. Thanks.

**Pull request merged**